### PR TITLE
KAFKA-14294: check whether a transaction is in flight before skipping a commit

### PR DESCRIPTION
**Pull request by ableegoldman**
Add a new `#transactionInFlight` API to the StreamsProducer to expose the flag of the same name, then check whether there is an open transaction when we determine whether or not to perform a commit in TaskExecutor. This is to avoid unnecessarily dropping out of the group on transaction timeout in the case a transaction was begun outside of regular processing, eg when a punctuator forwards records but there are no newly consumer records and thus no new offsets to commit

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java]: Wouldn't this lead to an NPE if there are no offsets to commit but a transaction is in flight?
See https://github.com/apache/kafka/blob/7387a1162533c52c83fb8899a31c505c42e0fe10/clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java#L752

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java]: nit: 
```suggestion
                if (!allOffsets.isEmpty() || taskManager.threadProducer().transactionInFlight()) {
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java:180]: nit:
Could you extract each branch to its own method so that for each processing mode we have a method that commits offsets or transaction. I think that would improve readability.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskExecutorTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskExecutorTest.java:70]:
```suggestion
        final StreamTask task = statelessTask(new TaskId(0, 0)).build();
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskExecutorTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskExecutorTest.java]: If my comment about the NPE above is correct, this verification needs to be changed.

**Review by cadonna — COMMENTED** [review]
@ableegoldman Thanks for the PR!

Here my feedback!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java]: ah, yes, we should pass in an empty map not `null`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java]: While breaking up each processing mode into its own separate branch as you requested, I decided to move the `allOffsets` map inside this if condition -- ie it still uses the `offsetsPerTask` map for the emptiness check. I'm assuming this is fine since it's just a nit but lmk if you still prefer to do the check on the `allOffsets` map instead

Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java]: Yes, moving `allOffsets` inside the branch is fine with me.
However, I actually proposed to extract each branch to a separate method, so that you have something like:

```java
if (executionMetadata.processingMode() == EXACTLY_ONCE_ALPHA) {
    commitForEosv1(...);
} else if (executionMetadata.processingMode() == EXACTLY_ONCE_V2) {
    commitForEosv2(...);
} else {
    commitForAlos(...);
}
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskExecutorTest.java
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskExecutorTest.java:70]: Do you have any concerns regarding using this helper methods to setup mocks?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: I would use a mock for `Tasks`.

**Review by cadonna — COMMENTED** [review]
@ableegoldman Thanks for the updates!

Here my feedback!

**Comment by ableegoldman** [comment]
All tests passing now, except for a handful of unrelated/flaky tests

Filed https://issues.apache.org/jira/browse/KAFKA-14384

@cadonna I'll wait for a final pass/approval from you before merging, or feel free to merge it yourself of course. I wonder if it's worth backporting to earlier versions? Maybe at least to 3.3 since we haven't done the bugfix release yet?

**Review by cadonna — APPROVED** [review]
@ableegoldman Thank you for the updates!

LGTM!

However, you missed some of my comments. Probably my bad since I directly replied to some comment instead of putting them into my last review. The comments are:
- https://github.com/apache/kafka/pull/12835#discussion_r1018815885
- https://github.com/apache/kafka/pull/12835#discussion_r1018813558

I could not find unit tests in `TaskExecutorTest` which verify that a transaction is committed when there are offsets. Also the ALOS does not seem to be tested.  

Do you plan a follow-up PR to address those things?

**Comment by cadonna** [comment]
Build failures are not related:

```
Build / JDK 11 and Scala 2.13 / kafka.api.ProducerIdExpirationTest.testProducerIdExpirationWithNoTransactions(String).quorum=zk
Build / JDK 11 and Scala 2.13 / kafka.network.DynamicConnectionQuotaTest.testDynamicListenerConnectionCreationRateQuota()
Build / JDK 11 and Scala 2.13 / org.apache.kafka.streams.integration.SelfJoinUpgradeIntegrationTest.shouldUpgradeWithTopologyOptimizationOff
Build / JDK 17 and Scala 2.13 / kafka.api.TransactionsExpirationTest.testTransactionAfterProducerIdExpires(String).quorum=zk
Build / JDK 17 and Scala 2.13 / org.apache.kafka.trogdor.coordinator.CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated()
```

**Comment by ableegoldman** [comment]
Sorry Bruno! I did miss those two comments it seems, apologies for that. I'll go back over your feedback and open a quick followup PR to make sure everything is addressed. Thanks!

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskExecutorTest.java
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskExecutorTest.java:70]: Nope, wasn't aware of this helper method's existence -- thanks for the tip

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java]: Ah I see. Will do

**Pull request merged**
